### PR TITLE
Classification forecasting for regression models

### LINKED ISSUE #1473: Timeseries classification
**Is your feature request related to a current problem? Please describe.**
Timeseries classification is not feasible in Darts, IoT has excellent data quality and interesting business cases, we've used Darts many times for regression achieving great results in short time, classification should be a feature in the roadmap since its becoming more important each day. 

**Describe proposed solution**
Being the head of the neural the network, I propose creating a subset of models that can tackle classification, RandomForest and Regresinos with scikit learn could raise a first a opportunity to have time series classification.

**Describe potential alternatives**
As a first approach Darts could take this as a regression problem with a distribution of probability between 0 and 1, and using a function like softmax to adjust to it to a class.

**Additional context**
-

### LINKED ISSUE #2315: Add Multi-Horizon Time Series Classification Support
Darts has established itself as a premier time series forecasting library. Adding multi-horizon time series classification support would solidify its position and significantly benefit researchers and practitioners alike. This feature aligns with previous discussions and interest, as seen in issue #1473.

Use Case Example:

Traffic congestion prediction: Classifying whether traffic congestion is likely to occur over multiple upcoming time intervals (e.g., next 15 minutes, next hour) based on historical traffic patterns.

Desired Functionality

- Input: Time series data and desired classification horizon.
- Output: Class labels for each time series segment across the specified horizon.

Potential Approaches

- Incorporate classifiers like LightGBMClassifier, enabling them to produce class label predictions for multiple future time steps within the Darts framework.

### PR DESCRIPTION
**Pull request by jonasblanc**
Checklist before merging this PR:
- [x] Mentioned all issues that this PR fixes or addresses.
- [x] Summarized the updates of this PR under **Summary**.
- [x] Added an entry under **Unreleased** in the [Changelog](../CHANGELOG.md).

Fixes #1473, Fixes #2315 

### Summary

Done:

- Add support for categorical target series  (refactor `RegressionModelWithCategoricalCovariates` into `RegressionModelWithCategoricalFeatures`)
- `_ForecastingClassifierMixin` for methods specific to forecasting classifier models.
- Support sklearn like classifier through `SklearnClassifierModel`
- Implement specific model for `XGBoost`, `LightGBM`, `Catboost` classifiers
- Support multi-labels classification through `MultiOutputClassifier` (implement `MultiOutputMixin` to extend both `MultiOutputClassifier` and `MultiOutputRegressor`)
- Extend tests to cover classification forecasting
- Implement `ClassProbabilityLikelihood` to predict class probabilities on `predict_likelihood_parameters`, make it default likelihood for classifier models.

TODO:
For future PR:
- Implement classification scores
- Write a notebook on classification forecasting

### CODE REVIEW

File: darts/utils/likelihood_models/classification.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+        if not hasattr(model, "classes_"):
+            raise_log(
+                ValueError(
+                    "The model must have a `classes_` attribute to fit the likelihood."
+                ),
+                logger,
+            )
+
+        classes = model.classes_
+        if type(classes) is not list:
```
Comment (dennisbader) [inline darts/utils/likelihood_models/classification.py:55]: Are we sure that all models only return lists and not any other sequence type (like numpy array or others)?

```suggestion
        if not isinstance(classes, list):
```

File: darts/utils/likelihood_models/classification.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+            classes = [classes]
+        self._classes = classes
+
+        unqiue_classes = {label for classes in classes for label in classes}
```
Comment (dennisbader) [inline darts/utils/likelihood_models/classification.py:59]:
```suggestion
        unique_classes = {label for classes in classes for label in classes}
```

File: darts/utils/likelihood_models/classification.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+        self._parameter_names = [
+            f"probability_class_{int(label)}" for label in unqiue_classes
```
Comment (dennisbader) [inline darts/utils/likelihood_models/classification.py:61]: Maybe just `p_{label}`? This will be added to the end of the component names. It should ideally be something very short.

For quantiles we have `q_{quantile}` for example.

```suggestion
            f"p_{int(label)}" for label in unqiue_classes
```

File: darts/utils/likelihood_models/classification.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+        ]
+        return self
+
+    def _estimator_predict(
+        self,
+        model,
+        x: np.ndarray,
+        **kwargs,
+    ) -> np.ndarray:
+        # TODO support multi-output with inequal number of classes
```
Comment (dennisbader) [inline darts/utils/likelihood_models/classification.py:71]: Yes, this would be required before merging :) 

also adding comments about the dimensions would be very helpful (similar to the other likelihoods)

File: darts/utils/likelihood_models/classification.py [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+            )
+        return preds.reshape(n_samples, self._n_outputs, -1)
+
+    def predict_likelihood_parameters(self, model_output: np.ndarray) -> np.ndarray:
+        """
+        Returns the distribution parameters as a array, extracted from the raw model outputs.
+        """
+        n_samples = model_output.shape[1]
+
+        # reshape to (n_series * n_samples, output_chunk_length, n_components)
```
Comment (dennisbader) [inline darts/utils/likelihood_models/classification.py:108]:
```suggestion
        # reshape to (n_series * n_samples, output_chunk_length, n_components * n_classes)
```

**Review by dennisbader — CHANGES_REQUESTED** [review]
Really nice job, thanks a lot @jonasblanc 🚀 

Looks like it's in pretty good shape already 👍 

I added a couple of suggestions. Really excited to see this in action in the future :)

File: darts/models/forecasting/catboost_model.py [hunk 6]
```diff
@@ -429,3 +456,243 @@ def _format_samples(
...
+            categorical_future_covariates=categorical_future_covariates,
+            categorical_static_covariates=categorical_static_covariates,
+            **kwargs,
+        )
+
+    def _create_model(self, **kwargs):
+        """Instanciate the underlying CatBoostClassifier model"""
+        return CatBoostClassifier(**kwargs)
+
+    def _set_likelihood(
```
[the remark this answers is not in the document]
Comment (jonasblanc) [reply darts/models/forecasting/catboost_model.py:646]: I did not include it so that it raises an error if the quantile is not None (which can only happen if we modify the code). Should I explicitly test it ?

File: darts/models/forecasting/regression_model.py [hunk 7]
```diff
@@ -1758,3 +1773,17 @@ def _format_samples(
...
+        """
+        Returns the name of the categorical features parameter from model's `fit` method .
+        """
+
+    @property
+    @abstractmethod
+    def _is_target_categorical(self) -> bool:
+        """ "
+        Returns if the target serie will be treated as categorical features when `lags` are provided.
+        """
```
[the remark this answers is not in the document]
Comment (jonasblanc) [reply darts/models/forecasting/regression_model.py:1789]: Yes _is_target_categorical is required because not all ModelType.FORECASTING_CLASSIFIER will treat target as categorical feature since they don't all have the capacity to do so. Only model inheriting from RegressionModelWithCategoricalFeatures have the capacity to do so. Which makes me realise this is not the correct place to define it. Will move it into ClassificationForecastingMixin.

[the remark this answers is not in the document]
Comment (jonasblanc) [reply darts/models/forecasting/regression_model.py:1789]: I did refactor _is_target_categorical into _is_categorical_features_supported in ClassificationForecastingMixin.
This is used to raise a warning if classifier does not support categorical feature so it's not the same as ModelType.FORECASTING_CLASSIFIER

[the remark this answers is not in the document]
Comment (dennisbader) [reply darts/models/forecasting/regression_model.py:1789]: You could just check if self is an instance of RegressionModelWithCategoricalFeatures :)

[the remark this answers is not in the document]
Comment (jonasblanc) [reply darts/models/forecasting/regression_model.py:1789]: Yes that's what _is_categorical_features_supported was doing. Since I was only using it at one place, I did remove the function and I test it directly

File: darts/utils/likelihood_models/classification.py (hunk 1 above)
Comment (jonasblanc) [reply darts/utils/likelihood_models/classification.py:55]: Actually all models following the sklearn interface should return an array. Only our MultiOutputClassifier returns a list of array. Since unifying our MultiOutputClassifier is not possible I treat all models as if they are MultiOutputClassifier with a single estimator. I'm not enforcing that the provided model should return an array on .classes_

File: darts/tests/models/forecasting/test_classifier_model.py [hunk 8]
```diff
@@ -0,0 +1,713 @@
...
+                assert any(record.levelname == "WARNING" for record in caplog.records)
+                assert any([
+                    message.startswith(
+                        "This model will treat the target `series` data/label as a numerical feature"
+                    )
+                    for message in caplog.messages
+                ])
+
+    @pytest.mark.parametrize("clf_params", process_model_list(classifiers))
+    def test_categorical_target_passed_to_fit_correctly(self, clf_params):
```
Comment (dennisbader) [inline darts/tests/models/forecasting/test_classifier_model.py:551]: Nice :) Could we try only running this for lgbm and catboost? Otherwise it will run a test for each model even though nothing is checked

File: darts/tests/models/forecasting/test_classifier_model.py [hunk 9]
```diff
@@ -0,0 +1,713 @@
...
+            model.predict(2, predict_likelihood_parameters=True)
+        print(err.value)
+        assert (
+            str(err.value) == "`predict_likelihood_parameters=True` is only"
+            " supported for probabilistic models fitted with a likelihood."
+        )
+
+        model_likelihood.fit(self.sine_univariate1_cat)
+        # model_likelihood has ClassProbability likelihood
+        probas = model_likelihood.predict(1, predict_likelihood_parameters=True)
```
Comment (dennisbader) [inline darts/tests/models/forecasting/test_classifier_model.py:705]: Nice :) 

We should add a couple more checks. You can use `darts.tests.models.forecasting.test_sf_models.TestSFModels.test_probabilistic_support` as a reference for what we should check (reproducibility, correct parameter names for parameter predictions, accuracy, sampled vs likelihood parameter vs median foreacst, ...).

File: darts/tests/models/forecasting/test_classifier_model.py
Comment (dennisbader) [inline darts/tests/models/forecasting/test_classifier_model.py:1]: Beautiful tests, congrats 🚀

**Review by dennisbader — CHANGES_REQUESTED** [review]
Very nice updates @jonasblanc 🚀 
I really like how it looks now :)

Had some more suggestions, but we're getting close to merging now 💯

File: darts/models/forecasting/catboost_model.py [hunk 10]
```diff
@@ -429,3 +456,243 @@ def _format_samples(
...
+class CatBoostClassifierModel(ClassificationForecastingMixin, CatBoostModel):
+    def __init__(
+        self,
+        lags=None,
+        lags_past_covariates=None,
+        lags_future_covariates=None,
+        output_chunk_length=1,
+        output_chunk_shift=0,
+        add_encoders=None,
+        likelihood=None,
```
[the remark this answers is not in the document]
Comment (jonasblanc) [reply darts/models/forecasting/catboost_model.py]: As agreed I did expose the likelihood in classifiers' constructor.

File: darts/utils/likelihood_models/sklearn.py [hunk 11]
```diff
@@ -412,59 +413,269 @@ def _get_median_prediction(self, model_output: np.ndarray) -> np.ndarray:
...
+    def predict_likelihood_parameters(self, model_output: np.ndarray) -> np.ndarray:
+        """
+        Returns the distribution parameters as a array, extracted from the raw model outputs.
+        """
+        # reshape to (n_series * n_samples, output_chunk_length, n_likelihood_parameters)
+        return model_output.transpose(1, 0, 2)
+
+    def _get_median_prediction(self, model_output: np.ndarray) -> np.ndarray:
+        """
+        Gets the median prediction per component extracted from the model output.
```
Comment (dennisbader) [inline darts/utils/likelihood_models/sklearn.py:603]:
```suggestion
        Gets the class label with highest predicted probability per component extracted
        from the model output.
```

File: darts/utils/likelihood_models/sklearn.py [hunk 12]
```diff
@@ -412,59 +413,269 @@ def _get_median_prediction(self, model_output: np.ndarray) -> np.ndarray:
...
+        """
+        # shape (output_chunk_length, n_series * n_samples, n_likelihood_parameters)
+        n_output, n_samples, _ = model_output.shape
+
+        n_component = len(self._classes) // self._n_outputs
+
+        # shape (output_chunk_length, n_series * n_samples, n_components)
+        preds = np.empty((n_output, n_samples, n_component), dtype=int)
+
+        for output_idx in range(n_output):
```
Comment (dennisbader) [inline darts/utils/likelihood_models/sklearn.py:559]: as you mentioned this can be quite inefficient for large number of samples (it is expected for example that users predict with `num_samples=1000`).

Can we improve this? Maybe we could pass the `num_samples` to this method and then instead of picking one label per sample, we pick `num_samples` per series?

**Review by dennisbader — CHANGES_REQUESTED** [review]
This is awesome @jonasblanc 🚀 
I only had minor suggestions and a comment about the way we sample now.

File: darts/utils/multioutput.py [hunk 13]
```diff
@@ -133,3 +137,37 @@ def supports_sample_weight(self) -> bool:
...
+
+class MultiOutputClassifier(MultiOutputMixin, sk_MultiOutputClassifier):
+    """
+    :class:`sklearn.utils.multioutput.MultiOutputClassifier` with a modified ``fit()`` method that also slices
+    validation data correctly. The validation data has to be passed as parameter ``eval_set`` in ``**fit_params``.
+    """
+
+    def fit(self, X, y, sample_weight=None, **fit_params):
+        super().fit(X=X, y=y, sample_weight=sample_weight, **fit_params)
+        self.classes_ = [estimator.classes_ for estimator in self.estimators_]
```
[the remark this answers is not in the document]
Comment (jonasblanc) [reply darts/utils/multioutput.py:157]: It does in the `fit` method which we don't call as we use the `_fit_estimator` instead.

File: darts/models/forecasting/xgboost.py [hunk 14]
```diff
@@ -343,6 +364,190 @@ def min_train_series_length(self) -> int:
...
+        >>> pred.values()
+        array([[0.],
+                [0.],
+                [0.],
+                [1.],
+                [0.],
+                [0.]])
+        """
+
+        # likelihood always set to ClassProbability as it's the only supported classifiaction likelihood
```
[the remark this answers is not in the document]
Comment (dennisbader) [reply darts/models/forecasting/xgboost.py:512]: I think it's not fixed yet

File: darts/utils/likelihood_models/sklearn.py [hunk 15]
```diff
@@ -412,59 +414,279 @@ def _get_median_prediction(self, model_output: np.ndarray) -> np.ndarray:
...
+            zip(
+                self._index_first_param_per_component[:-1],
+                self._index_first_param_per_component[1:],
+            )
+        ):
+            component_probabilities = model_output[:, :, component_start:component_end]
+            total_proba = component_probabilities.sum(
+                axis=2
+            )  # shape  (n_output, n_samples)
+            difference = np.ones((n_output, n_samples)) - total_proba
```
Comment (dennisbader) [inline darts/utils/likelihood_models/sklearn.py:578]:
```suggestion
            difference = 1. - total_proba
```

File: darts/utils/likelihood_models/sklearn.py [hunk 16]
```diff
@@ -412,59 +414,279 @@ def _get_median_prediction(self, model_output: np.ndarray) -> np.ndarray:
...
+            component_probabilities += (
+                difference[:, :, np.newaxis] / component_probabilities.shape[2]
+            )
+
+            # argmax stops at the first True
+            # first time the random sample is greater than cumulative probability
+            # [0.2, 0.1, 0.5, 0.2] -> [0.2, 0.3, 0.8, 1]
+            # random: 0.7 -> idx = 2, this outcomes has 0.5 probability of happening
+            sampled_idx = np.argmax(
+                self._rng.uniform(0, 1, size=(difference.shape))[:, :, np.newaxis]
```
Comment (dennisbader) [inline darts/utils/likelihood_models/sklearn.py:596]:
```suggestion
                self._rng.uniform(0, 1, size=difference.shape + (1,))
```

File: darts/utils/likelihood_models/sklearn.py [hunk 17]
```diff
@@ -412,59 +414,279 @@ def _get_median_prediction(self, model_output: np.ndarray) -> np.ndarray:
...
+                < np.cumsum(component_probabilities, axis=2),
```
Comment (dennisbader) [inline darts/utils/likelihood_models/sklearn.py:597]: I guess `<=` would include everything (also `p==1.0`)
```suggestion
                <= np.cumsum(component_probabilities, axis=2),
```

File: darts/utils/likelihood_models/sklearn.py [hunk 18]
```diff
@@ -412,59 +414,279 @@ def _get_median_prediction(self, model_output: np.ndarray) -> np.ndarray:
...
+            ] = params_proba
+
+        # shape (output_chunk_length, n_series * n_samples, n_likelihood_parameters)
+        # n_likelihood_parameters is the sum of the number of classes for each component
+        # n_likelihood_parameters are ordered as in component_names:
+        # [comp0_l1, comp0_l2, comp1_l1, comp1_l2, comp1_l3,...]
+        # should be accessed through _index_first_param_per_component (for comp1 => 2)
+        return class_proba
+
+    def sample(self, model_output: np.ndarray) -> np.ndarray:
```
Comment (dennisbader) [inline darts/utils/likelihood_models/sklearn.py:554]: love it 😍 🚀

**Review by dennisbader — CHANGES_REQUESTED** [review]
Very, very nice :) Love the new sampling algorithm 😍 

Just the last minor comments. Let me know once the remaining points are ready :)
